charmhelpers/contrib/openstack/context.py

```diff
@@ -2988,3 +2988,91 @@ def get_ovs_portdata(self, bond):
+        :param device_mappings_key: Config key for device mappings
+                                    (default: 'sriov-device-mappings')
+        :type device_mappings_key: Optional[str]
+        """
+        numvfs_key = numvfs_key or 'sriov-numvfs'
+        device_mappings_key = device_mappings_key or 'sriov-device-mappings'
+
+        devices = pci.PCINetDevices()
+        charm_config = config()
+        sriov_numvfs = charm_config.get(numvfs_key) or ''
+        sriov_device_mappings = charm_config.get(device_mappings_key) or ''
+
+        # create list of devices from sriov_device_mappings config option
+        sriov_mapped_devices = [
+            pair.split(':', 1)[1]
+            for pair in sriov_device_mappings.split()
+        ]
+
+        # create list of devices from sriov_numvfs config option
+        sriov_config_devices = {}
+        try:
+            for ifname, numvfs in (
+                    pair.split(':', 1)
+                    for pair in sriov_numvfs.split()):
+                sriov_config_devices.update({ifname: numvfs})
+        except ValueError:
+            # sriov_numvfs configuration option does not contain a list of
+            # devices
+            pass
+
+        self._map = {}
+        for device in devices.pci_devices:
+            if device and device.sriov:
+                if sriov_numvfs == 'auto':
+                    if (sriov_mapped_devices and
+                            device.interface_name not in sriov_mapped_devices):
+                        log('Excluding configuration of SR-IOV device {}'
+                            .format(device.interface_name),
+                            level=INFO)
+                        continue
+                    numvfs = int(device.sriov_totalvfs)
+                elif sriov_numvfs.isdigit():
+                    numvfs = min(int(sriov_numvfs), int(device.sriov_totalvfs))
+                else:
+                    if device.interface_name not in sriov_config_devices:
+                        log('Explicit SRIOV device configuration does not '
+                            'contain interface {}, skipping.'
+                            .format(device.interface_name),
+                            level=DEBUG)
+                        continue
+                    numvfs = min(
+                        int(sriov_config_devices[device.interface_name]),
+                        int(device.sriov_totalvfs))
+                    if numvfs < int(
+                            sriov_config_devices[device.interface_name]):
+                        log('Requested value for sriov-numfs ({}) too high '
+                            'for interface {}. Falling back to interface '
+                            'totalvfs value: {}'
+                            .format(sriov_config_devices[device.interface_name],
+                                    device.interface_name,
+                                    device.sriov_totalvfs),
+                            level=INFO)
```

Comment:
This is probably unmaintainable :)  I don't mean that to be unkind; it's just hard to follow.  Could it be split up into functions, or include comments explaining what the code is doing.  e.g. there's a number of uses of `min()` ... it's for a reason, but unless you really understand the domain, it's not clear.

Reply:
No worries, that is what reviews are for.

To be fair the code in question it is a lift from charm into charm-helpers exercise, and we need to carry forward the same overloading of some of the configuration options.  The code organization is flavored from it's origin in the charm it comes from.  I'll have a pass at dividing the individual handling into helper methods and add commentary to it.

Reply:
Ah, that makes sense.